src/basic/log.c

```diff
@@ -1478,3 +1516,47 @@ void log_setup(void) {
         if (log_on_console() && show_color < 0)
                 log_show_color(true);
 }
+
+LogContext* log_context_attach(LogContext *c) {
+        assert(c);
+        return LIST_PREPEND(ll, _log_context, c);
+}
+
+LogContext* log_context_detach(LogContext *c) {
+        if (!c)
+                return NULL;
+
+        LIST_REMOVE(ll, _log_context, c);
+        return NULL;
+}
+
+LogContext* log_context_new(char **fields) {
+        LogContext *c = new(LogContext, 1);
+        if (!c)
+                return NULL;
+
+        *c = (LogContext) {
+                .fields = fields,
+        };
+
+        return log_context_attach(c);
+}
+
+LogContext* log_context_free(LogContext *c) {
+        if (!c)
+                return NULL;
+
+        c->fields = strv_free(c->fields);
+        log_context_detach(c);
+        return mfree(c);
+}
+
+size_t log_context_num_fields(LogContext *context) {
+        size_t num_fields = 0;
+
+        LIST_FOREACH(ll, c, context)
+                STRV_FOREACH(s, c->fields)
+                        num_fields++;
+
+        return num_fields;
+}
```

Comment:
hmm, if we have large numbers of fields this likely will get quite expensive... a static counter next to the linked list might be more future proof there? This function is called on every log message after all